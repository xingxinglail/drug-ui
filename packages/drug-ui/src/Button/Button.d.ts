import * as React from 'react';
export declare type Variant = 'text' | 'outlined' | 'contained' | 'fab';
export declare type Color = 'default' | 'primary' | 'secondary' | 'inherit';
export declare type Size = 'small' | 'medium' | 'large';
export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
    variant?: Variant;
    color?: Color;
    size?: Size;
    disabled?: boolean;
    disableRipple?: boolean;
    fullWidth?: boolean;
    href?: string;
    round?: boolean;
    fab?: boolean;
    icon?: boolean;
    loading?: boolean;
    children?: React.ReactNode;
    component?: React.ElementType;
    ref?: React.Ref<HTMLButtonElement>;
}
declare const Button: React.FC<ButtonProps>;
export default Button;
