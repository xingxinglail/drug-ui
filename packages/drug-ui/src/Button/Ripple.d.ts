import * as React from 'react';
interface RippleProps {
    center?: boolean;
}
declare const Ripple: React.FC<RippleProps>;
export default Ripple;
