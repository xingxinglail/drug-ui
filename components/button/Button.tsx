import React, { FC, useRef } from 'react';
import { classes } from '../_util';
import { useRipple } from '../_use';

import './style/index.scss';

interface ButtonProps {
    disabled?: boolean;
    disableRipple?: boolean;
}

const Button: FC<ButtonProps> = ({ disabled = false, disableRipple = false, children }) => {
    const ref = useRef<HTMLButtonElement>(null);

    if (!disableRipple) useRipple({ ref });

    return (
        <button
            ref={ ref }
            disabled={ disabled }
            className={ classes('drug-button', 'drug-button-primary') }>
            { children }
        </button>
    );
};

export default Button;
