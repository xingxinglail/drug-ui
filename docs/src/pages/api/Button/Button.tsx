import * as React from 'react';
import MarkdownDocs from '../../../components/MarkdownDocs';
// @ts-ignore
import markdown from './button.md';

export default function Page () {
    return <MarkdownDocs markdown={ markdown } />;
}
