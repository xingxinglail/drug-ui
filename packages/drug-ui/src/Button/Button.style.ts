import { Style, Styles } from 'jss';
import { Theme } from '../styles';

const root: Style = {
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'center',
    position: 'relative',
    color: 'rgba(0, 0, 0, 0.87)',
    padding: [6, 16],
    fontSize: 14,
    minWidth: 64,
    boxSizing: 'border-box',
    transition: 'background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, border 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
    fontFamily: `"Roboto", "Helvetica", "Arial", sans-serif`,
    fontWeight: 500,
    lineHeight: 1.75,
    borderRadius: 4,
    letterSpacing: 0.4,
    verticalAlign: 'middle',
    border: 0,
    outline: 0,
    cursor: 'pointer',
    textDecoration: 'none',
    userSelect: 'none',
    '-webkit-appearance': 'none',
    '-webkit-tap-highlight-color': 'transparent',
    '&:hover': {
        textDecoration: 'none',
        backgroundColor: 'rgba(0, 0, 0, 0.08)'
    },
    '&$disabled': {
        cursor: 'default',
        pointerEvents: 'none',
        color: 'rgba(0, 0, 0, 0.26)'
    }
};

const contained: Style = () => ({
    boxShadow: '0 3px 1px -2px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 1px 5px 0px rgba(0, 0, 0, 0.12)',
    backgroundColor: '#e0e0e0',
    '&:hover': {
        boxShadow: '0 2px 4px -1px rgba(0, 0, 0, 0.2), 0px 4px 5px 0px rgba(0, 0, 0, 0.14), 0px 1px 10px 0px rgba(0, 0, 0, 0.12)',
        backgroundColor: '#d5d5d5'
    },
    '&$disabled': {
        boxShadow: 'none',
        backgroundColor: 'rgba(0, 0, 0, 0.12)'
    }
});

const containedPrimary: Style = (theme: Theme) => ({
    color: theme.palette.primary.contrastText,
    backgroundColor: theme.palette.primary.main,
    '&:hover': {
        backgroundColor: '#115293'
    }
});

const containedSecondary: Style = (theme: Theme) => ({
    color: theme.palette.secondary.contrastText,
    backgroundColor: theme.palette.secondary.main,
    '&:hover': {
        backgroundColor: '#9a0036'
    }
});

const text: Style = {
    padding: [6, 8]
};

const textPrimary: Style = {
    color: '#1976d2',
    '&:hover': {
        backgroundColor: 'rgba(25, 118, 210, 0.08)'
    }
};

const textSecondary: Style = {
    color: '#dc004e',
    '&:hover': {
        backgroundColor: 'rgba(220, 0, 78, 0.08)'
    }
};

const outlined: Style = {
    border: '1px solid rgba(0, 0, 0, 0.23)',
    padding: [5, 15],
    '&:hover': {
        backgroundColor: 'rgba(0, 0, 0, 0.08)'
    },
    '&$disabled': {
        border: '1px solid rgba(0, 0, 0, 0.26)'
    }
};

const outlinedPrimary: Style = {
    color: '#1976d2',
    border: '1px solid rgba(25, 118, 210, 0.5)',
    '&:hover': {
        border: '1px solid 1976d2',
        backgroundColor: 'rgba(25, 118, 210, 0.08)'
    }
};

const outlinedSecondary: Style = {
    color: '#dc004e',
    border: '1px solid rgba(220, 0, 78, 0.5)',
    '&:hover': {
        border: '1px solid rgb(220, 0, 78)',
        backgroundColor: 'rgba(220, 0, 78, 0.08)'
    }
};

const colorInherit: Style = {
    color: 'inherit',
    borderColor: 'currentColor'
};

const fab: Style = {
    minWidth: 'unset',
    width: 48,
    height: 48,
    borderRadius: '50%',
    padding: 0,
    boxShadow: '0 3px 1px -2px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 1px 5px 0px rgba(0, 0, 0, 0.12)',
    backgroundColor: '#e0e0e0',
    '&:hover': {
        boxShadow: '0 2px 4px -1px rgba(0, 0, 0, 0.2), 0px 4px 5px 0px rgba(0, 0, 0, 0.14), 0px 1px 10px 0px rgba(0, 0, 0, 0.12)',
        backgroundColor: '#d5d5d5'
    },
    '&$disabled': {
        boxShadow: 'none',
        backgroundColor: 'rgba(0, 0, 0, 0.12)'
    }
};

const fullWidth: Style = {
    width: '100%'
};

const disabled: Style = {};

const icon: Style = {
    ...fab,
    boxShadow: 'none',
    backgroundColor: null,
    '&:hover': null
};

export const styles = (theme: Theme): Styles => ({
    root: root,
    contained: contained(),
    containedPrimary: containedPrimary(theme),
    containedSecondary: containedSecondary(theme),
    text: text,
    textPrimary: textPrimary,
    textSecondary: textSecondary,
    outlined: outlined,
    outlinedPrimary: outlinedPrimary,
    outlinedSecondary: outlinedSecondary,
    colorInherit: colorInherit,
    fab: fab,
    fabPrimary: containedPrimary(theme),
    fabSecondary: containedSecondary(theme),
    icon,
    fullWidth,
    disabled
});
