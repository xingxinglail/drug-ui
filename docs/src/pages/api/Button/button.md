---
filename: /packages/drug-ui/src/Button/Button.js
---

# Button API

<p class="description">Button React组件的API文档。</p>

## Import

```js
import Button from '@drug-ui/core/Button';
// or
import { Button } from '@drug-ui/core';
```

## Props

| Name | Type | Default | Description |
|:-----|:-----|:--------|:------------|
| <span class="prop-name required">children&nbsp;*</span> | <span class="prop-type">node</span> |  | The content of the button. |
| <span class="prop-name">classes</span> | <span class="prop-type">object</span> |  | Override or extend the styles applied to the component. See [CSS API](#css) below for more details. |
| <span class="prop-name">color</span> | <span class="prop-type">'default'<br>&#124;&nbsp;'inherit'<br>&#124;&nbsp;'primary'<br>&#124;&nbsp;'secondary'</span> | <span class="prop-default">'default'</span> | The color of the component. It supports those theme colors that make sense for this component. |
| <span class="prop-name">component</span> | <span class="prop-type">elementType</span> | <span class="prop-default">'button'</span> | The component used for the root node. Either a string to use a DOM element or a component. |
| <span class="prop-name">disabled</span> | <span class="prop-type">bool</span> | <span class="prop-default">false</span> | If `true`, the button will be disabled. |
| <span class="prop-name">disableFocusRipple</span> | <span class="prop-type">bool</span> | <span class="prop-default">false</span> | If `true`, the  keyboard focus ripple will be disabled. `disableRipple` must also be true. |
| <span class="prop-name">disableRipple</span> | <span class="prop-type">bool</span> |  | If `true`, the ripple effect will be disabled.<br>⚠️ Without a ripple there is no styling for :focus-visible by default. Be sure to highlight the element by applying separate styles with the `focusVisibleClassName`. |
| <span class="prop-name">endIcon</span> | <span class="prop-type">node</span> |  | Element placed after the children. |
| <span class="prop-name">fullWidth</span> | <span class="prop-type">bool</span> | <span class="prop-default">false</span> | If `true`, the button will take up the full width of its container. |
| <span class="prop-name">href</span> | <span class="prop-type">string</span> |  | The URL to link to when the button is clicked. If defined, an `a` element will be used as the root node. |
| <span class="prop-name">size</span> | <span class="prop-type">'small'<br>&#124;&nbsp;'medium'<br>&#124;&nbsp;'large'</span> | <span class="prop-default">'medium'</span> | The size of the button. `small` is equivalent to the dense button styling. |
| <span class="prop-name">startIcon</span> | <span class="prop-type">node</span> |  | Element placed before the children. |
| <span class="prop-name">variant</span> | <span class="prop-type">'text'<br>&#124;&nbsp;'outlined'<br>&#124;&nbsp;'contained'</span> | <span class="prop-default">'text'</span> | The variant to use. |

`ref` 被转发到根元素.

## CSS

Style sheet name: `DuiButton`.

## Demos

- [Buttons](/components/buttons/)




