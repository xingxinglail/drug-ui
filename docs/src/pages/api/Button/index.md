---
filename: /packages/drug-ui/src/Button/Button.js
---

# Button API

<p class="description">Button React 组件的 API 文档。</p>

## Import

```js
import Button from '@drug-ui/core/Button';
// or
import { Button } from '@drug-ui/core';
```

## Props

| Name | Type | Default | Description |
|:-----|:-----|:--------|:------------|
| <span class="prop-name required">children&nbsp;*</span> | <span class="prop-type">node</span> |  | 按钮内容。 |
| <span class="prop-name">color</span> | <span class="prop-type">'default'<br>&#124;&nbsp;'inherit'<br>&#124;&nbsp;'primary'<br>&#124;&nbsp;'secondary'</span> | <span class="prop-default">'default'</span> | 组件颜色。 |
| <span class="prop-name">component</span> | <span class="prop-type">elementType</span> | <span class="prop-default">'button'</span> | 根节点组件。使用DOM元素的字符串或组件。 |
| <span class="prop-name">disabled</span> | <span class="prop-type">bool</span> | <span class="prop-default">false</span> | 禁用按钮。 |
| <span class="prop-name">fullWidth</span> | <span class="prop-type">bool</span> | <span class="prop-default">false</span> | 该按钮将占用其容器的全部宽度。 |
| <span class="prop-name">href</span> | <span class="prop-type">string</span> |  | 单击按钮时要链接到的URL。默认渲染 ```a``` 节点。 |
| <span class="prop-name">size</span> | <span class="prop-type">'small'<br>&#124;&nbsp;'medium'<br>&#124;&nbsp;'large'</span> | <span class="prop-default">'medium'</span> | 按钮大小。 |
| <span class="prop-name">variant</span> | <span class="prop-type">'text'<br>&#124;&nbsp;'outlined'<br>&#124;&nbsp;'contained'</span> | <span class="prop-default">'text'</span> | 多种风格。 |

`ref` 被转发到根元素.

## CSS

Style sheet name: `DuiButton`.

## Demos

- [Buttons](/drug-ui/components/Buttons)




