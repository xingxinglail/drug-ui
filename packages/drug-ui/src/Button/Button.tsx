import * as React from 'react';
import * as PropTypes from 'prop-types';
import { PropTypes as ComponentPropTypes } from '..';
import classnames from 'classnames';
import { createUseStyles } from '@drug-ui/styles';
import { Loading } from '@drug-ui/icons';
import { capitalize } from '../utils';
import { styles } from './Button.style';
import { Theme } from '../styles';
import ButtonBase, { ButtonBaseProps } from '../ButtonBase';

export type Variant = 'text' | 'outlined' | 'contained';

export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement>, ButtonBaseProps {
    variant?: Variant;
    color?: ComponentPropTypes.Color;
    size?: ComponentPropTypes.Size;
    fullWidth?: boolean;
    round?: boolean;
    loading?: boolean;
    ref?: React.Ref<HTMLButtonElement>
}

type ButtonClassProps =
    'root'
    | 'contained'
    | 'containedPrimary'
    | 'containedSecondary'
    | 'text'
    | 'textPrimary'
    | 'textSecondary'
    | 'outlined'
    | 'outlinedPrimary'
    | 'outlinedSecondary'
    | 'fab'
    | 'fabPrimary'
    | 'fabSecondary'
    | 'icon'
    | 'colorInherit'
    | 'fullWidth'
    | 'disabled'
    | 'containedSizeSmall'
    | 'containedSizeLarge'
    | 'textSizeSmall'
    | 'textSizeLarge'
    | 'outlinedSizeSmall'
    | 'outlinedSizeLarge'
    | 'round'
    | 'label'
    | 'loading';

const name = 'Button';

const useStyles = createUseStyles<Theme, ButtonClassProps>(styles, { name });

const Button: React.FC<ButtonProps> = React.forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
    const { className, variant = 'text', color = 'default', size = 'medium', disabled, component, fullWidth: defaultFullWidth, href, round, loading, children, ...rest } = props;
    const classes = useStyles();

    const classNames = classnames(
        classes.root,
        [classes[variant]],
        {
            [classes[`${ variant }${ capitalize(color) }`]]: color !== 'inherit' && color !== 'default',
            [classes[`${ variant }Size${ capitalize(size) }`]]: size !== 'medium',
            [classes.colorInherit]: color === 'inherit',
            [classes.fullWidth]: defaultFullWidth,
            [classes.round]: round,
            [classes.loading]: loading,
            [classes.disabled]: disabled
        }
    );

    const newChildren = React.Children.map(children, child => {
        const type = typeof child;
        if (type === 'string' || type === 'number') {
            return <span className={ classes.label }>{ child }</span>;
        }
        return child;
    });

    return (
        <ButtonBase
            className={ classNames }
            component={ component }
            ref={ ref }
            href={ href }
            disabled={ disabled }
            { ...rest }>
            <Loading/>
            { newChildren }
            {/*{ loading ? <Icons name="loading" className="loading"/> : null }*/ }
        </ButtonBase>
    );
});

Button.displayName = name;

Button.defaultProps = {
    variant: 'text',
    color: 'default',
    size: 'medium',
    fullWidth: false,
    loading: false,
    round: false
};

Button.propTypes = {
    variant: PropTypes.oneOf<Variant>(['text', 'outlined', 'contained']),
    color: PropTypes.oneOf<ComponentPropTypes.Color>(['default', 'primary', 'secondary', 'inherit']),
    size: PropTypes.oneOf<ComponentPropTypes.Size>(['small', 'medium', 'large']),
    loading: PropTypes.bool,
    fullWidth: PropTypes.bool,
    round: PropTypes.bool
};

export default Button;
