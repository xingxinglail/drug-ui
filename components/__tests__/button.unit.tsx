import * as renderer from 'react-test-renderer';
import * as React from 'react';
import Button from '../button';

describe('button', () => {
    it('是个 div', () => {
        const json = renderer.create(<Button/>).toJSON();
        expect(json).toMatchSnapshot();
    });
});
