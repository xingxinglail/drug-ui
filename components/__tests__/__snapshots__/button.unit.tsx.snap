// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button 是个 div 1`] = `
<div>
  Button
</div>
`;
